server/finder/handler/finder_handler.go

```diff
@@ -57,45 +57,56 @@ func (h *FinderHandler) Find(mhashes []multihash.Multihash) (*model.FindResponse
 		provResults := make([]model.ProviderResult, 0, len(values))
 		for j := range values {
 			iVal := values[j]
-			provID := iVal.ProviderID
-			pinfo := h.fetchProviderInfo(provID, iVal.ContextID, provInfos, true)
+			mainProvID := iVal.ProviderID
+			mainProvInfo := h.fetchProviderInfo(mainProvID, iVal.ContextID, provInfos, true)
 
-			if pinfo == nil {
+			if mainProvInfo == nil {
 				continue
 			}
 
 			// Adding the main provider
-			provResult := providerResultFromValue(provID, iVal.ContextID, iVal.MetadataBytes, pinfo.AddrInfo.Addrs)
-			provResults = append(provResults, provResult)
+			mainProvResult := providerResultFromValue(mainProvID, iVal.ContextID, iVal.MetadataBytes, mainProvInfo.AddrInfo.Addrs)
+			provResults = append(provResults, mainProvResult)
 
-			if pinfo.ExtendedProviders == nil {
+			if mainProvInfo.ExtendedProviders == nil {
 				continue
 			}
 
-			epRecord := pinfo.ExtendedProviders
+			epRecord := mainProvInfo.ExtendedProviders
 
-			// If override is set to true at the context level then the chain level EPs should be ignored for this context ID
-			override := false
+			contextualEpRecord, ok := epRecord.ContextualProviders[string(iVal.ContextID)]
 
-			// Adding context-level EPs if they exist
-			if contextualEpRecord, ok := epRecord.ContextualProviders[string(iVal.ContextID)]; ok {
-				override = contextualEpRecord.Override
-				for _, epInfo := range contextualEpRecord.Providers {
-					provResult := createExtendedProviderResult(epInfo, iVal)
-					provResults = append(provResults, *provResult)
+			// If override is set to true at the context level then the chain level EPs should be ignored for this context ID
+			if ok && !contextualEpRecord.Override {
+				for _, epInfo := range epRecord.Providers {
+					if epInfo.PeerID != mainProvID {
+						eProvResult := createExtendedProviderResult(epInfo, iVal)
+						provResults = append(provResults, *eProvResult)
+						continue
+					}
+
+					// updating main provider's metadata if it's not empty.
+					if len(epInfo.Metadata) > 0 {
+						(&provResults[0]).Metadata = epInfo.Metadata
```

Comment:
Addresses aren't included because they should match the ones specified in the AD's body and hence will be updated in the registry.